File: Csharp/chapter01/knock09.cs
/*
 * Chapter01, knock09
 * Task: Make Typoglycemia Sentence.
 */

using System;
using System.Linq;


public class TypoglycemiaSentence {
    static public void Main() {
        string sentence = "I couldn't believe that I could actually understand what I was reading : the phenomenal power of the human mind .";
       
        Console.WriteLine(Typoglycemia(sentence));
    }


    public static string Typoglycemia(string str, string d=" ") {
        string[] delimiter= new string[] {d};
        string[] words =  str.Split(delimiter, StringSplitOptions.None);
        string typoglycemiaSentence = "";
        for (int i=0; i < words.Length; i++) {
            int wordLength = words[i].Length;
            if (wordLength > 4) {
                typoglycemiaSentence += words[i][0] + ShuffleString(words[i].Substring(1,  wordLength - 2)) + words[i][wordLength-1]; 
            } else {
                typoglycemiaSentence += words[i];
            }
            if (i != words.Length-1)
                typoglycemiaSentence += " ";
        }
        return typoglycemiaSentence;
    }


    public static string ShuffleString(string str) {
        Random rand = new Random();
        return new string(str.ToCharArray().
                OrderBy(s => (rand.Next(2) % 2) == 0).ToArray());
    }
}

File: Ruby/Chapter01/knock09.rb
# typoglycemia program
# - if shuffle characters of a word without head and tail character,
#   the shuffled word is understanded.
# apply a sentence that is "I couldn't believe that I could actually understand what I was reading 
# : the phenomenal power of the human mind ."


def typoglycemia(word)
  if word.length < 3 
    return word
  end
  return word[0] + word.slice(1..-2).split('').shuffle.join + word[-1]
end


def sentence_typoglycemia(sentence)
  applyed_wordlist = []
  sentence.split(' ').each{|word| 
    applyed_wordlist.push(typoglycemia(word))
  }
  return applyed_wordlist.join(' ')
end

sentence = "I couldn't believe that I could actually understand what I was reading : the phenomenal power of the human mind ."

p sentence_typoglycemia(sentence)
